// autodll-webpack-plugin/lib/createConfig.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _webpack = require('webpack');

var _webpack2 = _interopRequireDefault(_webpack);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _paths = require('./paths');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var createConfig = function createConfig(_ref) {
  var hash = _ref.hash,
      filename = _ref.filename,
      entry = _ref.entry,
      plugins = _ref.plugins;

  var outputPath = _path2.default.join(_paths.cacheDir, hash);

  plugins = plugins || [];

  return {
    resolve: {
      extensions: ['.js', '.jsx']
    },
    entry: entry,
    output: {
      path: _path2.default.join(outputPath),
      filename: filename,
      library: '[name]_[hash]'
    },
    plugins: [new _webpack2.default.DllPlugin({
      path: _path2.default.join(outputPath, '[name].manifest.json'),
      name: '[name]_[hash]'
    })].concat(plugins)
  };
};

exports.default = createConfig;
