// autodll-webpack-plugin/lib/createSettings.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _utils = require('./utils');

var _createHash = require('./createHash');

var _createHash2 = _interopRequireDefault(_createHash);

var _normalizeEntry = require('./normalizeEntry');

var _normalizeEntry2 = _interopRequireDefault(_normalizeEntry);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var getInstanceId = function getInstanceId(index) {
  return `instance_${index}`;
};
var getEnv = function getEnv(env) {
  return env || process.env.NODE_ENV || 'development';
};

var createSettings = function createSettings(_ref) {
  var originalSettings = _ref.originalSettings,
      index = _ref.index;

  var entry = originalSettings.entry,
      env = originalSettings.env,
      otherSettings = _objectWithoutProperties(originalSettings, ['entry', 'env']);

  var defaults = {
    context: process.cwd(),
    path: '',
    entry: null,
    filename: '[name].js',
    inject: false,
    debug: false
  };

  var settings = (0, _utils.merge)(defaults, otherSettings, {
    entry: (0, _normalizeEntry2.default)(entry),
    id: getInstanceId(index),
    env: getEnv(env)
  });

  return (0, _utils.merge)(settings, {
    hash: (0, _createHash2.default)(settings)
  });
};

exports.default = createSettings;
