// rc-tween-one/es/plugin/PathPlugin.js
import _extends from 'babel-runtime/helpers/extends';
import { parsePath, getTransformValue } from '../util';
import { checkStyleName, getTransform } from 'style-utils';

function PathPlugin(target, vars) {
  this.target = target;
  var path = typeof vars === 'string' ? vars : vars.x || vars.y || vars.rotate;
  this.vars = vars;
  this.path = parsePath(path);
  this.start = {};
  this.pathLength = this.path.getTotalLength();
}

PathPlugin.prototype = {
  name: 'path',
  useStyle: 'transform',
  getPoint: function getPoint(offset) {
    var o = offset || 0;
    var p = this.pathLength * this.progress + o;
    return this.path.getPointAtLength(p);
  },
  getAnimStart: function getAnimStart(computedStyle) {
    var transform = getTransform(computedStyle[checkStyleName('transform')]);
    this.start = transform;
    this.data = _extends({}, transform);
  },
  setRatio: function setRatio(r, t) {
    this.progress = r;
    var p = this.getPoint();
    var p0 = this.getPoint(-1);
    var p1 = this.getPoint(1);
    if (typeof this.vars === 'string') {
      this.data.translateX = p.x + this.start.translateX;
      this.data.translateY = p.y + this.start.translateY;
      this.data.rotate = Math.atan2(p1.y - p0.y, p1.x - p0.x) * 180 / Math.PI;
    } else {
      this.data.translateX = this.vars.x ? p.x + this.start.translateX : this.data.translateX;
      this.data.translateY = this.vars.y ? p.y + this.start.translateY : this.data.translateY;
      this.data.rotate = this.vars.rotate ? Math.atan2(p1.y - p0.y, p1.x - p0.x) * 180 / Math.PI : this.data.rotate;
    }
    t.style.transform = getTransformValue(this.data);
  }
};

export default PathPlugin;
